clientv3/grpc_dial_test.go

```
@@ -0,0 +1,118 @@
+)
+
+// TestBalancerDoNotBlockOnClose ensures that balancer and grpc don't deadlock each other
+// due to rapid open/close conn. The deadlock causes balancer.Close() to block forever.
+// See issue: https://github.com/coreos/etcd/issues/7283 for more detail.
+func TestBalancerDoNotBlockOnClose(t *testing.T) {
+	defer testutil.AfterTest(t)
+
+	kcl := NewKillConnListener([]string{"127.0.0.1:8081", "127.0.0.1:8082"})
+	kcl.Start()
+	defer kcl.Close()
+
+	timer := time.After(10 * time.Second)
+	for {
+		select {
+		case <-timer:
+			return
+		default:
+		}
+		conn, err := grpc.Dial("127.0.0.1:8081", grpc.WithInsecure(),
+			grpc.WithBalancer(newSimpleBalancer([]string{"127.0.0.1:8081", "127.0.0.1:8082"})))
+		if err != nil {
+			t.Fatal(err)
+		}
+		// sleep a bit to allow connections to be established.
+		time.Sleep(100 * time.Millisecond)
+		closec := make(chan struct{})
+		// Close() might block on balancer.Close()
+		// if balancer and grpc deadlock each other.
+		go func() {
+			defer close(closec)
+			conn.Close()
+		}()
+		select {
+		case <-closec:
+		case <-time.After(time.Second):
+			t.Errorf("conn.Close() timeout")
+		}
+	}
+}
+
+// killConnListener listens incoming conn and kills it immediately.
+type killConnListener struct {
+	wg    sync.WaitGroup
+	eps   []string
+	donec chan struct{}
+
+	// mu protects lns
+	mu  sync.Mutex
+	lns []net.Listener
+}
+
+func NewKillConnListener(eps []string) *killConnListener {
+	return &killConnListener{eps: eps, donec: make(chan struct{})}
+}
+
+func (kcl *killConnListener) Start() {
+	for _, ep := range kcl.eps {
+		kcl.wg.Add(1)
+		go func(endpoint string) {
+			defer kcl.wg.Done()
+			kcl.listen(endpoint)
+		}(ep)
+	}
+}
+
+func (kcl *killConnListener) listen(url string) {
+	ln, _ := net.Listen("tcp", url)
+	kcl.mu.Lock()
+	kcl.lns = append(kcl.lns, ln)
+	kcl.mu.Unlock()
+	for {
+		select {
+		case <-kcl.donec:
+			return
+		default:
+		}
+		conn, err := ln.Accept()
+		if err != nil {
+			continue
```

do the listener close in this function and there's no need to ignore the error; this should panic